crates/noirc_frontend/src/hir/resolution/resolver.rs

```diff
@@ -610,14 +610,81 @@ impl<'a> Resolver<'a> {
             r#type: self.resolve_type(let_stmt.r#type),
             expression,
-        })
+        });
+        (hir_statement, self.errors)
+    }
+
+    pub fn resolve_global_let(
+        mut self,
+        let_stmt: crate::LetStatement,
+    ) -> (HirStatement, Vec<ResolverError>) {
+        match let_stmt.expression.kind.clone() {
+            // Ensure it is a literal.
+            ExpressionKind::Literal(literal) => {
+                // If it is a literal array, check the elements are literals themselves.
+                match literal {
+                    // Check each element of a standard array.
+                    Literal::Array(ArrayLiteral::Standard(elems)) => {
+                        let mut type_error = false;
+                        for e in elems.iter() {
+                            match e.kind {
+                                ExpressionKind::Literal(_) => (),
+                                // This element is not a literal.
+                                _ => {
+                                    type_error = true;
+                                    self.push_err(ResolverError::Expected {
+                                        span: let_stmt.expression.span,
+                                        expected: "global array elements can only be array, bool, integer or field literals".to_owned(), 
+                                        got: e.kind.to_string(),
+                                    });
+                                }
+                            }
+                        }
+                        if type_error {
+                            return (HirStatement::Error, self.errors);
+                        }
+                        self.resolve_literal(let_stmt)
+                    }
+                    // Just check the type of the repeated element.
+                    Literal::Array(ArrayLiteral::Repeated { repeated_element, length: _ }) => {
+                        match repeated_element.kind {
+                            ExpressionKind::Literal(_) => self.resolve_literal(let_stmt),
+                            // Repeated type is not a literal.
+                            _ => {
+                                self.push_err(ResolverError::Expected {
+                                    span: let_stmt.expression.span,
+                                    expected: "global array elements can only be array, bool, integer or field literals".to_owned(), 
+                                    got: repeated_element.kind.to_string(),
+                                });
+                                (HirStatement::Error, self.errors)
+                            }
+                        }
+                    }
+                    // Non-array literal is OK.
+                    _ => self.resolve_literal(let_stmt),
+                }
+            }
+            // Not a literal.
+            _ => {
+                self.push_err(ResolverError::Expected {
```

Comment:
sure! this is nicer. the new error messages are 

```
error: Invalid type found in global
   ┌─ /Users/michaelneuder/github/noir2/crates/nargo/tests/test_data/global_consts/src/main.nr:14:20
   │
14 │   global badGlobal = if M == 32 {
   │ ╭────────────────────'
15 │ │      constrain 1 == 1;
16 │ │ };
   │ ╰─' Valid types are: literal array, bool, integer or field

error: Invalid type found in global
   ┌─ /Users/michaelneuder/github/noir2/crates/nargo/tests/test_data/global_consts/src/main.nr:18:21
   │
18 │   global badGlobal2 = for i in 0..1 {
   │ ╭─────────────────────'
19 │ │      constrain i == 0;
20 │ │ };
   │ ╰─' Valid types are: literal array, bool, integer or field

error: Invalid type found in global
   ┌─ /Users/michaelneuder/github/noir2/crates/nargo/tests/test_data/global_consts/src/main.nr:23:12
   │
23 │ global A = [BAD, ELEMS];
   │            ------------ Valid types are: literal array, bool, integer or field

error: Invalid type found in global
   ┌─ /Users/michaelneuder/github/noir2/crates/nargo/tests/test_data/global_consts/src/main.nr:23:12
   │
23 │ global A = [BAD, ELEMS];
   │            ------------ Valid types are: literal array, bool, integer or field

error: aborting due to 4 previous errors
```

Reply:
https://github.com/noir-lang/noir/pull/826/commits/e89725c1b7789120f709373f42c5b14548755503 is the commit that adds this change

Reply:
Looks better, though we should avoid saying "invalid type" since these are not types in the type system. "Invalid expression" or "invalid expression kind" would be more accurate.

Reply:
nice! changed it to "expression kind"

```
nargo compile main
error: Invalid expression kind found in global
   ┌─ /Users/michaelneuder/github/noir2/crates/nargo/tests/test_data/global_consts/src/main.nr:14:20
   │
14 │   global badGlobal = if M == 32 {
   │ ╭────────────────────'
15 │ │      constrain 1 == 1;
16 │ │ };
   │ ╰─' Valid expression kinds are: literal array, bool, integer or field

error: Invalid expression kind found in global
   ┌─ /Users/michaelneuder/github/noir2/crates/nargo/tests/test_data/global_consts/src/main.nr:18:21
   │
18 │   global badGlobal2 = for i in 0..1 {
   │ ╭─────────────────────'
19 │ │      constrain i == 0;
20 │ │ };
   │ ╰─' Valid expression kinds are: literal array, bool, integer or field

error: Invalid expression kind found in global
   ┌─ /Users/michaelneuder/github/noir2/crates/nargo/tests/test_data/global_consts/src/main.nr:23:12
   │
23 │ global A = [BAD, ELEMS];
   │            ------------ Valid expression kinds are: literal array, bool, integer or field

error: Invalid expression kind found in global
   ┌─ /Users/michaelneuder/github/noir2/crates/nargo/tests/test_data/global_consts/src/main.nr:23:12
   │
23 │ global A = [BAD, ELEMS];
   │            ------------ Valid expression kinds are: literal array, bool, integer or field
```